README.md
# Hey it's Another Documentation Server! (hads)

[![NPM version](https://img.shields.io/npm/v/hads.svg)](https://www.npmjs.com/package/hads)
![Node version](https://img.shields.io/node/v/hads.svg)
![Downloads](https://img.shields.io/npm/dm/hads.svg)
[![License](https://img.shields.io/npm/l/hads.svg)](LICENSE)

### *The master of ~~hell~~ docs*

> **Hads** is a fast Node.js based web server allowing to browse, search and edit documentation written in
[Markdown](http://daringfireball.net/projects/markdown/).

![screenshot](https://cloud.githubusercontent.com/assets/593151/18916305/4dcdf5ba-8594-11e6-8eff-831153b784fb.png)

**Features**:

- No configuration needed
- Github-like presentation
- GFM ([Github Flavoured Markdown](https://guides.github.com/features/mastering-markdown/))
- Automatic indexation and search
- In-browser editor
- Table of contents using Markdown extension `[[toc]]`
- Navigation index using Markdown extension `[[index]]`
- Diagrams and flowcharts using [Mermaid](http://knsv.github.io/mermaid/) syntax
- Drag'n drop images
- 100% offline


## Usage

```bash
npm install -g hads
hads -o
```

Your browser will open `http://localhost:4040` and display your project documentation.

### Command-line options

```
Usage: hads [root dir] [options]

Options:
  -p, --port        Port number to listen on       [default: 4040]
  -h, --host        Host address to bind to        [default: "localhost"]
  -i, --images-dir  Directory to store images      [default: "images"]
  -o, --open        Open default browser on start
  --help            Show this help
```

If no root dir is specified, `./` will be used.

## Extras

### Home page
 
The server will automatically search for a file named `index.md`, `readme.md` or `README.md` on the specified
documentation root and will use it as your home page.

### Table of contents

The special text `[[toc]]` will be replaced by the table of contents of the markdown document, based on headings.

### Navigation index

The special text `[[index]]` will be replaced by the full navigation index of all markdown files found under the
specified *root dir*. File and folder names will be *humanized* for better readability.

It is particularly useful on the home page to provide an overview of the available documentation for your project.

### Diagrams and flowcharts

You can use the [Mermaid](http://knsv.github.io/mermaid/) syntax to insert diagrams and flowcharts directly in your
markdown, but using code blocks with the `mermaid` language specified, like this:

    ```mermaid
    graph TD;
        A-->B;
        A-->C;
        B-->D;
        C-->D;
    ```
    
## Updates

See changelog [here](CHANGELOG)


CHANGELOG
# 1.3.2

+ Added module export to allow server integration and extension

# 1.3.1

+ Added new mermaid neutral theme
- Fixed issue with dependencies path when installed as a local module

# 1.3.0

+ Added support for github-style markdown links (without extension)

# 1.2.2

+ Added image file selector button
- Fixed line endings on edited documents for unix platforms
- Fixed shape rendering for Mermaid graphs
- Fixed unwanted file dialog display
- Removed unwanted uploaded file previews

# 1.2.1

- Fixed edition of markdown files containing HTML tags

# 1.2.0

+ Added [[toc]] table of contents markdown extension
+ Added images drag'n drop support to editor
+ Added save (CTRL+S/CMD+S) keyboard shortcut to editor
+ Mermaid diagram code is now hidden during processing
- Removed specific rendering for images
- Small fixes and performance improvements

# 1.1.2

- Fixed issue on Windows

# 1.1.1

+ Added print-friendly styling
- Fixed issue when no home page was present

# 1.1.0

+ Added mermaid syntax support to generate diagrams
- Fixed [[index]] in code blocks
- Fixed search extracts

# 1.0.1

+ Adding new markdown files support
- Renamed project to "hads"


test/code.js
function hello(name) {
  console.log('Hello ' + name + '!');
}

hello('world');


mdds.js
'use strict';

let Promise = require('bluebird');
let fs = Promise.promisifyAll(require('fs'));
let mkdirpAsync = Promise.promisify(require('mkdirp'));
let path = require('path');
let optimist = require('optimist');
let express = require('express');
let bodyParser = require('body-parser');
let pkg = require('./package.json');
let Matcher = require('./lib/matcher.js');
let Renderer = require('./lib/renderer.js');
let Helpers = require('./lib/helpers.js');
let Indexer = require('./lib/indexer.js');

let args = optimist
  .usage(`\n${pkg.name} ${pkg.version}\nUsage: $0 [root dir] [options]`)
  .alias('p', 'port')
  .describe('p', 'Port number to listen on')
  .default('p', 4040)
  .alias('h', 'host')
  .describe('h', 'Host address to bind to')
  .default('h', 'localhost')
  .alias('o', 'open')
  .boolean('o')
  .describe('o', 'Open default browser on start')
  .describe('help', 'Show this help')
  .argv;

if (args.help || args._.length > 1) {
  optimist.showHelp(console.log);
  process.exit();
}

let docPath = args._[0] || './';
let rootPath = path.resolve(docPath);
let indexer = new Indexer(rootPath);
let renderer = new Renderer(indexer);
let app = express();

app.set('views', path.join(__dirname, '/views'));
app.set('view engine', 'pug');
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: true}));
app.use('/_mdds/', express.static(path.join(__dirname, '/public')));
app.use('/_mdds/highlight/', express.static(path.join(__dirname, 'node_modules/highlight.js/styles')));
app.use('/_mdds/octicons/', express.static(path.join(__dirname, 'node_modules/octicons/build/font')));
app.use('/_mdds/ace/', express.static(path.join(__dirname, 'node_modules/ace-builds/src-min/')));

const ROOT_FILES = ['index.md', 'README.md', 'readme.md'];
const STYLESHEETS = ['/highlight/github.css', '/octicons/octicons.css', '/css/github.css', '/css/style.css'];
const SCRIPTS = ['/ace/ace.js', '/js/client.js'];

app.get('*', (req, res, next) => {
  let route = Helpers.extractRoute(req.path);
  let query = req.query || {};
  let rootIndex = -1;
  let create = Helpers.hasQueryOption(query, 'create');
  let edit = Helpers.hasQueryOption(query, 'edit') || create;
  let filePath, icon, search, error, title;

  function tryProcessFile() {
    let contentPromise = null;
    filePath = path.join(rootPath, route);

    return fs.statAsync(filePath)
      .then((stat) => {
        search = query.search && query.search.length > 0 ? query.search.trim() : null;

        if (stat.isDirectory() && !search && !error) {
          if (!create) {
            // Try to find a root file
            route = path.join(route, ROOT_FILES[++rootIndex]);
            return tryProcessFile();
          } else {
            route = '/';
            title = 'Error';
            error = `Cannot create file \`${filePath}\``;
          }
        }

        if (error) {
          edit = false;
          contentPromise = Promise.resolve(Renderer.renderMarkdown(error));
          icon = 'octicon-alert';
        } else if (search) {
          contentPromise = renderer.renderSearch(query.search);
          icon = 'octicon-search';
        } else if (Helpers.hasQueryOption(query, 'raw')) {
          // Access raw content: images, code, etc
          return res.sendFile(filePath);
        } else if (Matcher.isMarkdown(filePath)) {
          contentPromise = edit ? renderer.renderRaw(filePath) : renderer.renderFile(filePath);
          icon = 'octicon-file';
        } else if (Matcher.isImage(filePath)) {
          contentPromise = renderer.renderImageFile(route);
          icon = 'octicon-file-media';
        } else if (Matcher.isSourceCode(filePath)) {
          contentPromise = renderer.renderSourceCode(filePath, path.extname(filePath).replace('.', ''));
          icon = 'octicon-file-code';
        }

        if (!title) {
          title = search ? renderer.searchResults : path.basename(filePath);
        }

        if (contentPromise) {
          return contentPromise.then((content) => {
            res.render(edit ? 'edit' : 'file', {
              title: title,
              route: route,
              icon: icon,
              search: search,
              content: content,
              styles: STYLESHEETS,
              scripts: SCRIPTS,
              pkg: pkg
            });
          });
        } else {
          next();
        }
      })
      .catch(() => {
        if (create) {
          let fixedRoute = Helpers.ensureMarkdownExtension(route);
          if (fixedRoute !== route) {
            return res.redirect(fixedRoute + '?create=1');
          }

          return mkdirpAsync(path.dirname(filePath))
            .then(() => fs.writeFileAsync(filePath, ''))
            .then(() => indexer.updateIndexForFile(filePath))
            .catch((e) => {
              console.error(e);
              title = 'Error';
              error = `Cannot create file \`${filePath}\``;
              route = '/';
            })
            .then(tryProcessFile);
        } else if (rootIndex !== -1 && rootIndex < ROOT_FILES.length - 1) {
          route = path.join(path.dirname(route), ROOT_FILES[++rootIndex]);
        } else {
          route = '/';
          title = '404 Error';
          error = '## File not found ¯\\\\\\_(◕\\_\\_◕)_/¯\n> *There\'s a glitch in the matrix...*';
        }
        return tryProcessFile();
      });
  }

  tryProcessFile();
});

app.post('*', (req, res, next) => {
  let route = Helpers.extractRoute(req.path);
  let filePath = path.join(rootPath, route);

  fs.statAsync(filePath)
    .then((stat) => {
      if (stat.isFile() && req.body.content) {
        return fs.writeFileAsync(filePath, req.body.content);
      }
    })
    .then(() => {
      indexer.updateIndexForFile(filePath);
      return renderer.renderFile(filePath);
    })
    .then((content) => {
      res.render('file', {
        title: path.basename(filePath),
        route: route,
        icon: 'octicon-file',
        content: content,
        styles: STYLESHEETS,
        scripts: SCRIPTS,
        pkg: pkg
      });
    })
    .catch(() => {
      next();
    })
});

indexer.indexFiles().then(() => {
  app.listen(args.port, args.host, () => {
    let serverUrl = `http://${args.host}:${args.port}`;
    console.log(`${pkg.name} ${pkg.version} serving at ${serverUrl} (press CTRL+C to exit)`);

    if (args.open) {
      require('open')(serverUrl);
    }
  });
});


lib/indexer.js
'use strict';

let Promise = require('bluebird');
let path = require('path');
let fs = Promise.promisifyAll(require('fs'));
let readdirAsync = Promise.promisify(require('recursive-readdir'));
let elasticlunr = require('elasticlunr');
let Matcher = require('./matcher');

const EXCLUSIONS = ['node_modules'];

class Indexer {

  constructor(basePath) {
    this.basePath = basePath;
    this.index = elasticlunr(function () {
      this.setRef('file');
      this.addField('filename');
      this.addField('content');
    });
    this.files = [];
  }

  indexFiles() {
    return readdirAsync(this.basePath, EXCLUSIONS).then((files) => {
      files = files.filter((file) => Matcher.isMarkdown(file));
      return Promise.all(files.map(this.updateIndexForFile.bind(this)));
    });
  }

  updateIndexForFile(file) {
    let filePath = path.relative(this.basePath, file);
    this.files.push(filePath);

    return fs.readFileAsync(path.join(this.basePath, filePath), 'utf8').then((content) => {
      this.index[this.index.documentStore.hasDoc(filePath) ? 'updateDoc' : 'addDoc']({
        file: filePath,
        filename: path.basename(filePath, path.extname(filePath)),
        content: content
      })
    });
  }

  getContent(file) {
    return this.index.documentStore.getDoc(file).content;
  }

  getFiles() {
    return this.files;
  }

  search(string) {
    return this.index.search(string, { expand: true });
  }

}

module.exports = Indexer;


lib/matcher.js
'use strict';

let path = require('path');
let _ = require('lodash');

const MARKDOWN_EXTENSIONS = ['md', 'mkdn', 'mdown', 'markdown'];
const IMAGES_EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'svg', 'webp'];
const SOURCE_CODE_EXTENSIONS = ['js', 'json', 'ts', 'coffee', 'css', 'scss', 'sass', 'less', 'stylus', 'html', 'jade',
  'pug', 'sh', 'txt'];

class Matcher {

  static isMarkdown(file) {
    return Matcher._hasExtension(file, MARKDOWN_EXTENSIONS);
  }

  static isImage(file) {
    return Matcher._hasExtension(file, IMAGES_EXTENSIONS);
  }

  static isSourceCode(file) {
    return Matcher._hasExtension(file, SOURCE_CODE_EXTENSIONS);
  }

  static _hasExtension(file, extensions) {
    return _.includes(extensions, path.extname(file).replace('.', ''));
  }

}

Matcher.MARKDOWN_EXTENSIONS = MARKDOWN_EXTENSIONS;

module.exports = Matcher;


lib/renderer.js
'use strict';

let Promise = require('bluebird');
let path = require('path');
let fs = Promise.promisifyAll(require('fs'));
let marked = require('marked');
let highlight = require('highlight.js');
let removeMd = require('remove-markdown');
let humanize = require('string-humanize');

const SEARCH_EXTRACT_LENGTH = 400;
const SEARCH_RESULTS_MAX = 10;

class Renderer {

  constructor(indexer) {
    this.indexer = indexer;
    this.searchResults = null;
  }

  renderRaw(file) {
    return fs.readFileAsync(file, 'utf8');
  }

  renderFile(file) {
    return this.renderRaw(file).then((content) => this.renderMarkdown(content));
  }

  renderSourceCode(file, lang) {
    return this.renderRaw(file).then((content) => this.renderMarkdown(`\`\`\`${lang}\n${content}\n\`\`\``));
  }

  renderSearch(search) {
    let results = this.indexer.search(search);
    let total = results.length;
    let content = total ? '' : 'No results.';

    this.searchResults = 'Search results (';

    if (total > SEARCH_RESULTS_MAX) {
      results.splice(SEARCH_RESULTS_MAX);
      this.searchResults += `first ${SEARCH_RESULTS_MAX} of `;
    }

    this.searchResults += `${total})`;

    results.forEach((result) => {
      let extract = removeMd(this.indexer.getContent(result.ref));
      extract = extract.replace(/\s/g, ' ').replace(/`/g, '');
      extract = extract.replace(/\[\[index]]/g, '&#91;&#91;index]]');

      if (extract.length > SEARCH_EXTRACT_LENGTH) {
        extract = extract.substr(0, SEARCH_EXTRACT_LENGTH) + ' [...]';
      }

      content += `[${result.ref}](${result.ref})\n> ${extract}\n\n`;
    });

    return Promise.resolve(this.renderMarkdown(content));
  }

  renderIndex() {
    let files = this.indexer.getFiles();
    let nav = {};

    files.forEach((file) => {
      let dir = path.dirname(file);
      let components = dir.split(path.sep);
      let name = path.basename(file);
      let parent = nav;

      if (components[0] === '.') {
        components.splice(0, 1);
      }

      components.forEach((component) => {
        let current = parent[component] || {};
        parent[component] = current;
        parent = current;
      });

      parent[name] = file;
    });

    let content = Renderer.renderIndexLevel(nav, 0);
    return this.renderMarkdown(content);
  }

  renderTableOfContents(content) {
    let toc = '';
    let renderer = new marked.Renderer();
    renderer.heading = (text, level) => {
      let slug = text.toLowerCase().replace(/[^\w]+/g, '-');
      toc += '  '.repeat(level) + `- [${text}](#${slug})\n`;
    };
    marked(content, {renderer: renderer});
    return this.renderMarkdown(toc);
  }

  renderMarkdown(content) {
    let renderer = new marked.Renderer();
    renderer.code = (code, language) => {
      if (language === 'mermaid') {
        return `<p class="mermaid">${code}</p>`;
      } else {
        return marked.Renderer.prototype.code.call(renderer, code, language);
      }
    };
    renderer.paragraph = (text) => {
      text = text.replace(/^\[\[toc]]/img, () => this.renderTableOfContents(content));
      text = text.replace(/^\[\[index]]/img, () => this.renderIndex());
      return marked.Renderer.prototype.paragraph.call(renderer, text);
    };

    return marked(content, {
      renderer: renderer,
      gfm: true,
      tables: true,
      smartLists: true,
      breaks: false,
      smartypants: true,
      highlight: (code, lang) => lang && lang !== 'no-highlight' ? highlight.highlight(lang, code, true).value : code
    });
  }

  static renderIndexLevel(index, level) {
    let content = '';
    let indent = '  '.repeat(level);
    let keys = Object.keys(index).sort((a, b) => {
      let aType = typeof index[a];
      let bType = typeof index[b];
      if (aType === bType) {
        return a.localeCompare(b);
      } else if (aType === 'string') {
        // Display files before folders
        return -1;
      }
      return 1;
    });

    keys.forEach((key) => {
      let value = index[key];
      content += indent;

      if (typeof value === 'string') {
        key = path.basename(key, path.extname(key));
        content += `- [${humanize(key)}](/${value})\n`;
      } else {
        content += `- ${humanize(key)}\n`;
        content += Renderer.renderIndexLevel(value, level + 1);
      }
    });

    return content;
  }

}

module.exports = Renderer;


public/js/client.js
window.hads = (function () {
  'use strict';

  var editor = null;
  var byId = document.getElementById.bind(document);
  var hads = {
    save: function () {
      if (editor) {
        byId('content').value = editor.getValue();
        byId('form').submit();
      }
    },
    showAdd: function () {
      byId('add-modal').className = '';
      byId('add').focus();
    },
    closeAdd: function () {
      byId('add-modal').className = 'hidden';
    },
    add: function (file) {
      this.closeAdd();
      window.location = '/' + file + '?create=1';
    },
    uploadImage: function() {
      document.getElementsByClassName('dz-hidden-input')[0].click();
    }
  };

  function getUploadMessage(file) {
    return '![Uploading ' + file + '...]()';
  }

  window.onload = function () {
    var editorElem = byId('editor');
    if (editorElem) {
      editor = ace.edit('editor');
      editor.setOptions({
        useSoftTabs: true,
        tabSize: 2,
        mode: 'ace/mode/markdown',
        theme: 'ace/theme/clouds',
        highlightActiveLine: false,
        wrap: true
      });
      var session = editor.getSession();
      var count = session.getLength();
      editor.focus();
      editor.gotoLine(count, session.getLine(count - 1).length);
      editor.commands.addCommand({
        name: 'save',
        bindKey: {
          win: 'Ctrl-S',
          mac: 'Command-S'
        },
        exec: hads.save,
        readOnly: true
      });

      var route = byId('route').value;
      Dropzone.autoDiscover = false;
      new Dropzone('.ace_scroller', {
        url: '/_hads/upload?route=' + encodeURI(route),
        acceptedFiles: 'image/*',
        maxFilesize: 10,  // 10 MB
        addedfile: function() {},
        init: function () {
          this.on('processing', function (file) {
            var pos = session.getSelection().getCursor();
            session.insert(pos, getUploadMessage(file.name));
          });
          this.on('success', function (file, path) {
            editor.replace('![' + file.name + '](' + path + ')', {
              needle: getUploadMessage(file.name)
            });
          });
          this.on('error', function (file) {
            editor.replace('![Error!]()', {
              needle: getUploadMessage(file.name)
            });
          });
        }
      });
    }

    document.addEventListener('click', function (e) {
      if (e.target.id === 'add-modal')
        window.hads.closeAdd();
    }, false);
    document.addEventListener('keydown', function (e) {
      if (e.keyCode === 27)
        window.hads.closeAdd();
    }, false);
  };

  return hads;
})();

hads.js
'use strict';

const Promise = require('bluebird');
const fs = Promise.promisifyAll(require('fs'));
const mkdirpAsync = Promise.promisify(require('mkdirp'));
const os = require('os');
const path = require('path');
const optimist = require('optimist');
const express = require('express');
const multer = require('multer');
const bodyParser = require('body-parser');
const shortId = require('shortid');
const pkg = require('./package.json');
const Matcher = require('./lib/matcher.js');
const Renderer = require('./lib/renderer.js');
const Helpers = require('./lib/helpers.js');
const Indexer = require('./lib/indexer.js');

let args = optimist
  .usage(`\n${pkg.name} ${pkg.version}\nUsage: $0 [root dir] [options]`)
  .alias('p', 'port')
  .describe('p', 'Port number to listen on')
  .default('p', 4040)
  .alias('h', 'host')
  .describe('h', 'Host address to bind to')
  .default('h', 'localhost')
  .alias('i', 'images-dir')
  .describe('i', 'Directory to store images')
  .default('i', 'images')
  .alias('o', 'open')
  .boolean('o')
  .describe('o', 'Open default browser on start')
  .describe('help', 'Show this help')
  .argv;

if (args.help || args._.length > 1) {
  optimist.showHelp(console.log);
  process.exit();
}

// Find node_modules base path
let modulesBasePath = require.resolve('highlight.js');
modulesBasePath = modulesBasePath.substr(0, modulesBasePath.lastIndexOf('node_modules'));

let docPath = args._[0] || './';
let rootPath = path.resolve(docPath);
let imagesPath = path.join(rootPath, Helpers.sanitizePath(args.i));
let indexer = new Indexer(rootPath);
let renderer = new Renderer(indexer);
let app = express();

app.set('views', path.join(__dirname, '/views'));
app.set('view engine', 'pug');
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: true}));
app.use('/_hads/', express.static(path.join(__dirname, '/public')));
app.use('/_hads/highlight/', express.static(path.join(modulesBasePath, 'node_modules/highlight.js/styles')));
app.use('/_hads/octicons/', express.static(path.join(modulesBasePath, 'node_modules/octicons/build/font')));
app.use('/_hads/ace/', express.static(path.join(modulesBasePath, 'node_modules/ace-builds/src-min/')));
app.use('/_hads/mermaid/', express.static(path.join(modulesBasePath, 'node_modules/mermaid/dist/')));
app.use('/_hads/dropzone/', express.static(path.join(modulesBasePath, 'node_modules/dropzone/dist/min/')));

const ROOT_FILES = ['index.md', 'README.md', 'readme.md'];
const STYLESHEETS = ['/highlight/github.css', '/octicons/octicons.css', '/css/github.css', '/css/style.css',
  '/css/mermaid.neutral.css'];
const SCRIPTS = ['/ace/ace.js', '/mermaid/mermaid.min.js', '/dropzone/dropzone.min.js', '/js/client.js'];

app.get('*', (req, res, next) => {
  let route = Helpers.extractRoute(req.path);
  let query = req.query || {};
  let rootIndex = -1, mdIndex = -1;
  let create = Helpers.hasQueryOption(query, 'create');
  let edit = Helpers.hasQueryOption(query, 'edit') || create;
  let filePath, icon, search, error, title, contentPromise;

  function renderPage() {
    if (error) {
      edit = false;
      contentPromise = Promise.resolve(renderer.renderMarkdown(error));
      icon = 'octicon-alert';
    } else if (search) {
      contentPromise = renderer.renderSearch(query.search);
      icon = 'octicon-search';
    } else if (Helpers.hasQueryOption(query, 'raw') || Matcher.isImage(filePath)) {
      return res.sendFile(filePath);
    } else if (Matcher.isMarkdown(filePath)) {
      contentPromise = edit ? renderer.renderRaw(filePath) : renderer.renderFile(filePath);
      icon = 'octicon-file';
    } else if (Matcher.isSourceCode(filePath)) {
      contentPromise = renderer.renderSourceCode(filePath, path.extname(filePath).replace('.', ''));
      icon = 'octicon-file-code';
    }

    if (!title) {
      title = search ? renderer.searchResults : path.basename(filePath);
    }

    if (contentPromise) {
      return contentPromise.then((content) => {
        res.render(edit ? 'edit' : 'file', {
          title: title,
          route: route,
          icon: icon,
          search: search,
          content: content,
          styles: STYLESHEETS,
          scripts: SCRIPTS,
          pkg: pkg
        });
      });
    } else {
      next();
    }
  }

  function tryProcessFile() {
    contentPromise = null;
    filePath = path.join(rootPath, route);

    return fs.statAsync(filePath)
      .then((stat) => {
        search = query.search && query.search.length > 0 ? query.search.trim() : null;

        if (stat.isDirectory() && !search && !error) {
          if (!create) {
            // Try to find a root file
            route = path.join(route, ROOT_FILES[++rootIndex]);
            return tryProcessFile();
          } else {
            route = '/';
            title = 'Error';
            error = `Cannot create file \`${filePath}\``;
          }
        }

        return renderPage();
      })
      .catch(() => {
        if (create) {
          let fixedRoute = Helpers.ensureMarkdownExtension(route);
          if (fixedRoute !== route) {
            return res.redirect(fixedRoute + '?create=1');
          }

          return mkdirpAsync(path.dirname(filePath))
            .then(() => fs.writeFileAsync(filePath, ''))
            .then(() => indexer.updateIndexForFile(filePath))
            .then(tryProcessFile)
            .catch((e) => {
              console.error(e);
              title = 'Error';
              error = `Cannot create file \`${filePath}\``;
              route = '/';
              return renderPage();
            });
        } else if (rootIndex !== -1 && rootIndex < ROOT_FILES.length - 1) {
          route = path.join(path.dirname(route), ROOT_FILES[++rootIndex]);
          return tryProcessFile();
        } else if (rootIndex === -1 && path.basename(route) !== '' && (path.extname(route) === '' || mdIndex > -1) &&
            mdIndex < Matcher.MARKDOWN_EXTENSIONS.length - 1) {
          // Maybe it's a github-style link without extension, let's try adding one
          let extension = Matcher.MARKDOWN_EXTENSIONS[++mdIndex];
          route = path.join(path.dirname(route), `${path.basename(route, path.extname(route))}.${extension}`);
          return tryProcessFile();
        } else {
          if (path.dirname(route) === path.sep && rootIndex === ROOT_FILES.length - 1) {
            error = '## No home page (╥﹏╥)\nDo you want to create an [index.md](/index.md?create=1) or ' +
              '[readme.md](/readme.md?create=1) file perhaps?'
          } else {
            error = '## File not found ¯\\\\\\_(◕\\_\\_◕)_/¯\n> *There\'s a glitch in the matrix...*';
          }
          title = '404 Error';
          route = '/';

          return renderPage();
        }
      });
  }

  tryProcessFile();
});

app.post('*', (req, res, next) => {
  let route = Helpers.extractRoute(req.path);
  let filePath = path.join(rootPath, route);

  fs.statAsync(filePath)
    .then((stat) => {
      let fileContent = req.body.content;
      if (stat.isFile() && fileContent) {
        if (process.platform !== 'win32') {
          // www-form-urlencoded data always use CRLF line endings, so this is a quick fix
          fileContent = fileContent.replace(/\r\n/g, '\n');
        }
        return fs.writeFileAsync(filePath, fileContent);
      }
    })
    .then(() => {
      indexer.updateIndexForFile(filePath);
      return renderer.renderFile(filePath);
    })
    .then((content) => {
      res.render('file', {
        title: path.basename(filePath),
        route: route,
        icon: 'octicon-file',
        content: content,
        styles: STYLESHEETS,
        scripts: SCRIPTS,
        pkg: pkg
      });
    })
    .catch(() => {
      next();
    })
});

app.post('/_hads/upload', [multer({
  storage: multer.diskStorage({
    destination: (req, file, cb) => { cb(null, imagesPath); },
    filename: (req, file, cb) => {
      mkdirpAsync(imagesPath).then(() => {
        cb(null, shortId.generate() + path.extname(file.originalname))
      });
    }
  }),
  onFileUploadStart: (file) => !file.mimetype.match(/^image\//),
  limits: {
    fileSize: 1024 * 1024 * 10   // 10 MB
  }
}).single('file'), (req, res) => {
  res.json(path.sep + path.relative(rootPath, req.file.path));
}]);

indexer.indexFiles().then(() => {
  app.listen(args.port, args.host, () => {
    let serverUrl = `http://${args.host}:${args.port}`;
    console.log(`${pkg.name} ${pkg.version} serving at ${serverUrl} (press CTRL+C to exit)`);

    if (args.open) {
      require('open')(serverUrl);
    }
  });
});

exports = app;
